File: android_native/app/src/main/kotlin/com/afluffywaffle/layuv/ai/OpenAiCompatibleProvider.kt
package com.afluffywaffle.layuv.ai

import android.util.Base64
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.UnknownHostException

/**
 * OpenAI-compatible chat-completions client (`POST {baseUrl}/chat/completions`,
 * Bearer auth, SSE) — covers OpenAI, Google Gemini (via its OpenAI-compat
 * endpoint), and local servers (Ollama / LM Studio / llama.cpp / vLLM, or a Mac
 * "brain"). The sole AI client in Layuv — streams for timeout safety, renders
 * once, and logs minimally (never the key or the request body / manuscript).
 */
class OpenAiCompatibleProvider(
    private val baseUrl: String,
    private val model: String,
    // Cloud endpoints (Gemini) require a key; a user's local server usually has none,
    // so the custom provider passes false and we send without an Authorization header.
    private val requireKey: Boolean = true,
) : AiProvider {

    override fun send(apiKey: String, messages: List<AiMessage>): AiResult {
        if (requireKey && apiKey.isBlank()) return AiResult.Error("Set your API key in AI settings first.")
        if (baseUrl.isBlank()) return AiResult.Error("Set the server address (base URL) in AI settings first.")
        val endpoint = baseUrl.trimEnd('/') + "/chat/completions"
        // Refuse plain HTTP to the public internet — only private/trusted hosts (see CleartextPolicy).
        CleartextPolicy.cleartextError(endpoint)?.let { return AiResult.Error(it) }
        var conn: HttpURLConnection? = null
        return try {
            conn = (URL(endpoint).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = true
                connectTimeout = 30_000
                readTimeout = 180_000
                if (apiKey.isNotBlank()) setRequestProperty("Authorization", "Bearer $apiKey")
                setRequestProperty("content-type", "application/json")
                setRequestProperty("accept", "text/event-stream")
            }
            conn.outputStream.use { it.write(buildBody(messages).toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            Log.i(TAG, "request sent; response $code")
            if (code != 200) mapHttpError(code, readErr(conn)) else parseStream(conn)
        } catch (e: UnknownHostException) {
            AiResult.Error("Couldn't reach the server — check the address and your connection.")
        } catch (e: IOException) {
            AiResult.Error("Network error — please try again.")
        } catch (e: Exception) {
            Log.w(TAG, "ai request failed: ${e.javaClass.simpleName}")
            AiResult.Error("Something went wrong. Please try again.")
        } finally {
            conn?.disconnect()
        }
    }

    private fun buildBody(messages: List<AiMessage>): String {
        val arr = JSONArray()
        for (m in messages) {
            val msg = JSONObject().put("role", m.role)
            if (m.images.isEmpty()) {
                msg.put("content", m.text)
            } else {
                // Multimodal: text first, then each handwritten-note PNG as a data-URI image part.
                val content = JSONArray().put(JSONObject().put("type", "text").put("text", m.text))
                for (img in m.images) {
                    val uri = "data:image/png;base64," + Base64.encodeToString(img, Base64.NO_WRAP)
                    content.put(
                        JSONObject().put("type", "image_url")
                            .put("image_url", JSONObject().put("url", uri)),
                    )
                }
                msg.put("content", content)
            }
            arr.put(msg)
        }
        return JSONObject()
            .put("model", model)
            .put("max_tokens", MAX_TOKENS)
            .put("stream", true)
            .put("messages", arr)
            .toString()
    }

    private fun parseStream(conn: HttpURLConnection): AiResult {
        val text = StringBuilder()
        var finish: String? = null
        BufferedReader(InputStreamReader(conn.inputStream, Charsets.UTF_8)).use { r ->
            while (true) {
                val line = r.readLine() ?: break
                if (!line.startsWith("data:")) continue
                val payload = line.substring(5).trim()
                if (payload.isEmpty() || payload == "[DONE]") continue
                val obj = try { JSONObject(payload) } catch (e: Exception) { continue }
                val c0 = obj.optJSONArray("choices")?.optJSONObject(0) ?: continue
                c0.optJSONObject("delta")?.optString("content")?.let { if (it.isNotEmpty()) text.append(it) }
                val fr = c0.optString("finish_reason")
                if (fr.isNotEmpty() && fr != "null") finish = fr
            }
        }
        val out = text.toString()
        return if (out.isBlank()) AiResult.Error("The AI returned an empty response. Please try again.")
        else AiResult.Ok(out, truncated = finish == "length")
    }

    private fun readErr(conn: HttpURLConnection): String? = try {
        conn.errorStream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }
    } catch (e: Exception) {
        null
    }

    private fun mapHttpError(code: Int, body: String?): AiResult.Error {
        val apiMsg = body?.let {
            try { JSONObject(it).optJSONObject("error")?.optString("message")?.takeIf { m -> m.isNotBlank() } }
            catch (e: Exception) { null }
        }
        val msg = when (code) {
            401, 403 -> "Your API key was rejected. Check it in AI settings."
            404 -> apiMsg ?: "Model or endpoint not found — check the model name."
            429 -> "Rate limited — please wait a moment and try again."
            in 500..599 -> "The AI server is busy right now. Please try again shortly."
            else -> apiMsg ?: "The request was rejected (HTTP $code)."
        }
        return AiResult.Error(msg)
    }

    companion object {
        private const val TAG = "AI"
        private const val MAX_TOKENS = 16000
    }
}

File: macos_native/LeamhApp/LeamhApp/Shared/OpenAiCompatibleProvider.swift
import Foundation

/// Errors surfaced by the AI provider. All cases carry a user-readable description.
enum AiError: LocalizedError {
    case notConfigured
    case cleartextBlocked(String)
    case http(Int, String)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "No AI endpoint configured. Set one in AI Settings."
        case .cleartextBlocked(let host):
            return "Plain HTTP is not allowed for \"\(host)\". Use HTTPS or a private/local address."
        case .http(let code, let msg):
            return "Server error \(code): \(msg)"
        case .networkError(let e):
            return e.localizedDescription
        }
    }
}

/// Streams completions from an OpenAI-compatible SSE endpoint (`POST /chat/completions`).
/// Supports multimodal requests (inline base64 PNG images as `image_url` content items).
enum OpenAiCompatibleProvider {

    /// Sentinel yielded as the last stream element when `finish_reason == "length"`.
    static let truncatedMarker = "__TRUNCATED__"

    /// Client-error codes a text-only model returns when it can't accept `image_url`
    /// parts. Mirror of the Kotlin twin's `TEXT_ONLY_REJECT_CODES`.
    private static let textOnlyRejectCodes: Set<Int> = [400, 404, 415, 422]

    /// Substituted for the dropped ink images on a text-only retry.
    /// Mirror of `AskAiPanel.INK_OMITTED_NOTE`.
    static let inkOmittedNote =
        "\n\n[Note: this model can't read images, so the handwritten note(s) "
        + "referenced above were not included. Work from the text only.]"

    /// Returns an `AsyncThrowingStream` of partial token strings.
    /// The stream ends normally on `[DONE]`. If truncated it first yields `truncatedMarker`.
    /// Call-site should cancel the stream on user dismissal; `onTermination` cancels the task.
    static func stream(
        turns: [AiTurn],
        images: [Data] = [],
        settings: AiProviderSettings = .shared
    ) throws -> AsyncThrowingStream<String, Error> {
        guard settings.isConfigured else { throw AiError.notConfigured }
        let url = try buildURL(settings)
        guard CleartextPolicy.isAllowed(url) else {
            throw AiError.cleartextBlocked(url.host ?? url.absoluteString)
        }
        let request = try makeRequest(
            url: url, turns: turns, images: images, settings: settings)
        // Prepared up front so the retry path can't throw inside the stream task.
        let textOnlyRequest: URLRequest? = images.isEmpty ? nil : try makeRequest(
            url: url, turns: turns, images: [], settings: settings, inkOmitted: true)

        return AsyncThrowingStream { continuation in
            let marker = truncatedMarker
            let task = Task {
                do {
                    var (bytes, response) = try await URLSession.shared.bytes(for: request)
                    if let http = response as? HTTPURLResponse,
                       !(200..<300).contains(http.statusCode) {
                        // Read the error body and surface the server's error.message
                        // when present (parity with the Kotlin twin's mapHttpError),
                        // falling back to the generic per-code message.
                        var data = Data()
                        for try await b in bytes { data.append(b) }
                        // A client error on a request that carried ink-note images is
                        // almost always a text-only model rejecting the image_url parts
                        // (Kotlin: mapHttpError → AiResult.NeedsTextOnlyRetry). Re-send
                        // ONCE without them, with a placeholder in the text.
                        var didRetry = false
                        if let retry = textOnlyRequest,
                           Self.textOnlyRejectCodes.contains(http.statusCode) {
                            (bytes, response) = try await URLSession.shared.bytes(for: retry)
                            didRetry = true
                        }
                        if let http2 = response as? HTTPURLResponse,
                           !(200..<300).contains(http2.statusCode) {
                            // On a retry, `data` holds the FIRST response's body; the
                            // second response's body is still unread, so drain it.
                            var retryData = data
                            if didRetry {
                                retryData = Data()
                                for try await b in bytes { retryData.append(b) }
                            }
                            let serverMsg = Self.serverErrorMessage(retryData)
                            throw AiError.http(http2.statusCode,
                                               serverMsg ?? httpMessage(http2.statusCode))
                        }
                    }
                    var truncated = false
                    for try await line in bytes.lines {
                        guard line.hasPrefix("data: ") else { continue }
                        let payload = String(line.dropFirst(6))
                        if payload == "[DONE]" { break }
                        if let token = parseToken(payload, truncated: &truncated) {
                            continuation.yield(token)
                        }
                        if truncated { break }
                    }
                    if truncated { continuation.yield(marker) }
                    continuation.finish()
                } catch is CancellationError {
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: - Private helpers

    private static func buildURL(_ settings: AiProviderSettings) throws -> URL {
        var base = settings.baseUrl.trimmingCharacters(in: .whitespaces)
        while base.hasSuffix("/") { base = String(base.dropLast()) }
        let path = base + "/chat/completions"
        guard let url = URL(string: path) else { throw AiError.networkError(URLError(.badURL)) }
        return url
    }

    /// Builds the POST request. `inkOmitted` marks the text-only retry: the ink
    /// images are gone, so the first user turn carries `inkOmittedNote` instead.
    private static func makeRequest(
        url: URL,
        turns: [AiTurn],
        images: [Data],
        settings: AiProviderSettings,
        inkOmitted: Bool = false
    ) throws -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod  = "POST"
        request.setValue("application/json",  forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        if let key = SecureKeyStore.read(key: SecureKeyStore.apiKeyName), !key.isEmpty {
            request.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try buildBody(
            turns: turns, images: images, settings: settings, inkOmitted: inkOmitted)
        return request
    }

    private static func buildBody(
        turns: [AiTurn],
        images: [Data],
        settings: AiProviderSettings,
        inkOmitted: Bool = false
    ) throws -> Data {
        var messages: [[String: Any]] = []

        for (i, turn) in turns.enumerated() {
            var turn = turn
            // Load-bearing: `turns` carries NO system message on either platform, so
            // index 0 is the user seed. If a system turn is ever prepended here, both
            // this note AND the image attachment below must be re-indexed.
            if inkOmitted && turn.role == AiTurn.roleUser && i == 0 {
                turn = AiTurn(role: turn.role,
                              text: turn.text + inkOmittedNote,
                              truncated: turn.truncated)
            }
            // Images are attached to the FIRST user message as content items.
            if turn.role == AiTurn.roleUser && i == 0 && !images.isEmpty {
                var content: [[String: Any]] = [["type": "text", "text": turn.text]]
                for img in images {
                    content.append([
                        "type": "image_url",
                        "image_url": ["url": "data:image/png;base64,\(img.base64EncodedString())"],
                    ])
                }
                messages.append(["role": turn.role, "content": content])
            } else {
                messages.append(["role": turn.role, "content": turn.text])
            }
        }

        let payload: [String: Any] = [
            "model":      settings.model,
            "messages":   messages,
            "max_tokens": 16000,
            "stream":     true,
        ]
        return try JSONSerialization.data(withJSONObject: payload)
    }

    /// Extracts the content delta from a single SSE `data:` payload.
    /// Sets `truncated = true` if `finish_reason == "length"`.
    private static func parseToken(_ json: String, truncated: inout Bool) -> String? {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = obj["choices"] as? [[String: Any]],
              let first = choices.first else { return nil }
        if let reason = first["finish_reason"] as? String, reason == "length" {
            truncated = true
        }
        guard let delta  = first["delta"] as? [String: Any],
              let text   = delta["content"] as? String,
              !text.isEmpty else { return nil }
        return text
    }

    /// Extracts `{"error":{"message":...}}` from an error-response body, if present.
    private static func serverErrorMessage(_ data: Data) -> String? {
        guard !data.isEmpty,
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let error = obj["error"] as? [String: Any],
              let message = error["message"] as? String,
              !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        else { return nil }
        return message
    }

    private static func httpMessage(_ code: Int) -> String {
        switch code {
        case 401: return "Unauthorized — check your API key in AI Settings."
        case 403: return "Forbidden."
        case 404: return "Endpoint not found — check the base URL in AI Settings."
        case 429: return "Rate limited — wait a moment and try again."
        case 500...599: return "The server returned an error."
        default: return HTTPURLResponse.localizedString(forStatusCode: code)
        }
    }
}
